keps/sig-node/3953-node-resource-hot-plug/README.md

```diff
@@ -0,0 +1,955 @@
+
+#### Story 5: Avoid Workload Disruption
+
+As a Cluster administrator, I expect my existing workloads to function without having to undergo a disruption which is induced during capacity addition followed by a node/kubelet restart to
+detect the change in compute capacity, which can bring in further complications.
+
+### Notes/Constraints/Caveats (Optional)
+
+### Risks and Mitigations
+
+- Change in Swap limit:
+    - The formula to calculate the swap limit is `(<containerMemoryRequest>/<nodeTotalMemory>)*<totalPodsSwapAvailable>`
+    - With change in nodeTotalMemory and totalPodsSwapAvailable post up-scale, The existing swap limit may not be inline with the
+      actual swap limit for existing pods.
+        - This can be mitigated by recalculating the swap limit for the existing pods. However, there can be an associated overhead for
+          recalculating the scores.
+
+- Change in OOMScoreAdjust value:
+    - The formula to calculate OOMScoreAdjust is `1000 - (1000*containerMemReq)/memoryCapacity`
+    - With change in memoryCapacity post up-scale, The existing OOMScoreAdjust may not be inline with the
+      actual OOMScoreAdjust for existing pods.
+    - Its not recommended to update the OOMScoreAdjust of a running container as OOMScoreAdjust value is set for init process(pid 1) which is 
+      responsible for running all other container's processes.
+    - When we update OOMScoreAdjust for a running container, its set for container init only, and possibly processes which will be started later and
+      running won't get the OOMScoreAdjust new value.
+        - This can be mitigated by updating the OOMScoreAdj formula to not consider current memory value, hence the new OOMScoreAdj formula looks like this
+            `min(0, 1000 - (1000*containerMemoryRequest)/initialMemoryCapacity)`
+
+- Post up-scale any failure in resync of Resource managers may be lead to incorrect or rejected allocation, which can lead to underperformed or rejected workload.
+  - To mitigate the risks adequate tests should be added to avoid the scenarios where failure to resync resource managers can occur.
+
+- Lack of coordination about change in resource availability across kubelet/runtime/NRI plugins.
+  - The runtime/NRI plugins should be updated to react to change in resource information on the node.
+
+- Kubelet missing on processing hotplug instance(s)
+  - Kubelet observes the underlying node for any hotplug of resources as and when generated, 
+    this ensures that the capacity is updated in set intervals and can technically not miss to update the actual capacity obtained from cAdvisor.
+
+- Handling downsize events
+  - Though, there is no support through this KEP to handle an event of node-downsize, it's the onus of the cluster administrator to resize responsibly to avoid disruption as it lies out of the kubernetes realm.
+  - However, in a situation of downsize an error mode is returned by the kubelet and the node is marked as `NotReady`.
```

Comment:
If an admin downsizes then increases resources after, does the node return to ready or stay in not ready until it's restarted? I feel the latter but should probably expand

Reply:
The current PoC implementation returns the node to Ready state if the resources (or more) become available again. Node is put into NotReady states but other parts of kubelet are not aware of the change (e.g. resource managers are not re-initialized etc). This kind of matches the existing behavior of kubelet not being aware of changes in node capacity.

Example of resizing a resource capacity: 10 (Ready) -> 5 (NotReady) -> 20 (Ready) -> 10 (NotReady) -> 15 (NotReady) -> 20 (Ready).